### PR TITLE
Integer -> Union{Integer,StaticInt}

### PR DESCRIPTION
**Pull request by Tokazama**

### CODE REVIEW

File: src/llvm_intrin/intrin_funcs.jl [hunk 1]
```diff
@@ -115,13 +115,13 @@ if Base.libllvm_version ≥ v"12"
   @inline vmin(v1::Vec{W,<:Unsigned}, v2::Vec{W,<:Signed}) where {W} =
     vifelse(v1 < v2, v1, v2)
 else
-  @inline vmax(v1::Vec{W,<:Integer}, v2::Vec{W,<:Integer}) where {W} =
+  @inline vmax(v1::Vec{W,<:Union{Integer,StaticInt}}, v2::Vec{W,<:Union{Integer,StaticInt}}) where {W} =
```
Comment (chriselrod) [inline src/llvm_intrin/intrin_funcs.jl:118]: I think the `Vec`s may be able to remain `<:Integer` and disallow `StaticInt`, but that's not currently the case.
I'm unsure if they're ever used. So leaving it as a Union like the PR currently is is probably fine.

But do make sure they don't hit `llvmcall_expr`.

File: src/llvm_intrin/intrin_funcs.jl [hunk 2]
```diff
@@ -623,11 +623,11 @@ end
...
-  ("vector.reduce.add", :vsum, :Integer),
-  ("vector.reduce.mul", :vprod, :Integer),
-  ("vector.reduce.and", :vall, :Integer),
-  ("vector.reduce.or", :vany, :Integer),
-  ("vector.reduce.xor", :vxorreduce, :Integer),
+  ("vector.reduce.add", :vsum, :(Union{Integer,StaticInt})),
+  ("vector.reduce.mul", :vprod, :(Union{Integer,StaticInt})),
+  ("vector.reduce.and", :vall, :(Union{Integer,StaticInt})),
+  ("vector.reduce.or", :vany, :(Union{Integer,StaticInt})),
+  ("vector.reduce.xor", :vxorreduce, :(Union{Integer,StaticInt})),
```
Comment (chriselrod) [inline src/llvm_intrin/intrin_funcs.jl:630]: These should all remain `Integer`.

File: src/special/misc.jl [hunk 3]
```diff
@@ -154,7 +154,7 @@ end
 @inline Base.FastMath.hypot_fast(x::AbstractSIMD, y::AbstractSIMD) =
   sqrt(Base.FastMath.add_fast(Base.FastMath.mul_fast(x, x), Base.FastMath.mul_fast(y, y)))
 
-@inline Base.clamp(x::AbstractSIMD{<:Any,<:Integer}, r::AbstractUnitRange{<:Integer}) =
+@inline Base.clamp(x::AbstractSIMD{<:Any,<:Union{Integer,StaticInt}}, r::AbstractUnitRange{<:Union{Integer,StaticInt}}) =
```
Comment (chriselrod) [inline src/special/misc.jl:157]: Could perhaps use
https://github.com/JuliaSIMD/VectorizationBase.jl/blob/c9194d38f3b2bfdf8a1dbc1e97cc0e8562831424/src/VectorizationBase.jl#L195
for some of this?

File: src/llvm_intrin/intrin_funcs.jl [hunk 4]
```diff
@@ -727,9 +727,9 @@ function count_zeros_func(W, I, op, tf = 1)
   rettypexpr = :(_Vec{$W,$I})
   llvmcall_expr(decl, instrs, rettypexpr, :(Tuple{$rettypexpr}), vtyp, [vtyp], [:(data(v))])
 end
-# @generated Base.abs(v::Vec{W,I}) where {W, I <: Integer} = count_zeros_func(W, I, "abs", 0)
-@generated vleading_zeros(v::Vec{W,I}) where {W,I<:Integer} = count_zeros_func(W, I, "ctlz")
-@generated vtrailing_zeros(v::Vec{W,I}) where {W,I<:Integer} =
+# @generated Base.abs(v::Vec{W,I}) where {W, I <: Union{Integer,StaticInt}} = count_zeros_func(W, I, "abs", 0)
+@generated vleading_zeros(v::Vec{W,I}) where {W,I<:Union{Integer,StaticInt}} = count_zeros_func(W, I, "ctlz")
```
Comment (chriselrod) [inline src/llvm_intrin/intrin_funcs.jl:731]:
```suggestion
@generated vleading_zeros(v::Vec{W,I}) where {W,I<:IntegerTypesHW} = count_zeros_func(W, I, "ctlz")
```

File: src/llvm_intrin/intrin_funcs.jl [hunk 5]
```diff
@@ -727,9 +727,9 @@ function count_zeros_func(W, I, op, tf = 1)
...
+@generated vtrailing_zeros(v::Vec{W,I}) where {W,I<:Union{Integer,StaticInt}} =
```
Comment (chriselrod) [inline src/llvm_intrin/intrin_funcs.jl:732]:
```suggestion
@generated vtrailing_zeros(v::Vec{W,I}) where {W,I<:IntegerTypesHW} =
```

File: src/llvm_intrin/masks.jl [hunk 6]
```diff
@@ -347,7 +347,7 @@ end
 #     EVLMask{$W,$M}($(typemax(M)) >>> ($(M(8sizeof(M))-1) - evl), evl + one(evl))
 #   end
 # end
-@generated function _mask_bzhi(::Union{Val{W},StaticInt{W}}, l::I) where {W,I<:Integer}
+@generated function _mask_bzhi(::Union{Val{W},StaticInt{W}}, l::I) where {W,I<:Union{Integer,StaticInt}}
```
Comment (chriselrod) [inline src/llvm_intrin/masks.jl:350]: I should perhaps add a TODO to add a separate dispatch for creating masks from `StaticInt`s.
I should change these lines:
https://github.com/JuliaSIMD/VectorizationBase.jl/blob/c9194d38f3b2bfdf8a1dbc1e97cc0e8562831424/src/llvm_intrin/masks.jl#L469-L472
They should all be redirecting to
`_mask_shift(::StaticInt{W}, l, ::False)`
which is the only mask method that doesn't use `llvmcall`, and would thus let Julia constprop them.
Or I could do `@generated`.

File: src/llvm_intrin/masks.jl [hunk 7]
```diff
@@ -507,7 +507,7 @@ end
 # @inline tounsigned(m::Vec{W,Bool}) where {W} = getfield(tomask(m), :u)
 @inline tounsigned(v) = getfield(tomask(v), :u)
 
-@generated function vrem(m::Mask{W,U}, ::Type{I}) where {W,U,I<:Integer}
+@generated function vrem(m::Mask{W,U}, ::Type{I}) where {W,U,I<:Union{Integer,StaticInt}}
```
Comment (chriselrod) [inline src/llvm_intrin/masks.jl:510]:
```suggestion
@generated function vrem(m::Mask{W,U}, ::Type{I}) where {W,U,I<:IntegerTypesHW}
```

File: src/llvm_intrin/masks.jl [hunk 8]
```diff
@@ -607,7 +607,7 @@ for (f, cond) ∈ [(:veq, "eq"), (:vne, "ne")]
   @eval @generated function $f(
     v1::Vec{W,T1},
     v2::Vec{W,T2},
-  ) where {W,T1<:Integer,T2<:Integer}
+  ) where {W,T1<:Union{Integer,StaticInt},T2<:Union{Integer,StaticInt}}
```
Comment (chriselrod) [inline src/llvm_intrin/masks.jl:610]:
```suggestion
  ) where {W,T1<:IntegerTypesHW,T2<:IntegerTypesHW}
```

**Comment by Tokazama** [comment]
Sorry, I'm realizing my mass find and replace did a lot of dumb changes.

File: src/llvm_intrin/masks.jl (hunk 6 above)
Comment (Tokazama) [reply src/llvm_intrin/masks.jl:350]: Do you want me to just add a todo note here?

**Comment by chriselrod** [comment]
I should add downstream tests.
```julia
(T, #= /home/chriselrod/.julia/dev/LoopVectorization/test/miscellaneous.jl:967 =# @__LINE__()) = (Float32, 967)
Miscellaneous: Error During Test at /home/chriselrod/.julia/dev/LoopVectorization/test/miscellaneous.jl:10
  Got exception outside of a @test
  MethodError: no method matching (LayoutPointers.FastRange{Int32})(::Int64, ::Static.StaticInt{1})
  Closest candidates are:
    (LayoutPointers.FastRange{T})(::F, ::S, ::O) where {T, F, S, O} at ~/.julia/dev/LayoutPointers/src/stridedpointers.jl:229
  Stacktrace:
    [1] memory_reference
      @ ~/.julia/dev/LayoutPointers/src/stridedpointers.jl:238 [inlined]
    [2] stridedpointer_preserve
      @ ~/.julia/dev/LayoutPointers/src/stridedpointers.jl:97 [inlined]
    [3] (::var"#sumprodavx#57")(x::UnitRange{Int32})
      @ Main ~/.julia/dev/LoopVectorization/test/miscellaneous.jl:580
```
and
```julia
typeinf_edge at ./compiler/typeinfer.jl:895
abstract_call_method at ./compiler/abstractinterpretation.jl:601
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:161
abstract_call_known at ./compiler/abstractinterpretation.jl:1695
abstract_call at ./compiler/abstractinterpretation.jl:1767
abstract_call at ./compiler/abstractinterpretation.jl:1732
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1888
typeinf_local at ./compiler/abstractinterpretation.jl:2341
typeinf_nocycle at ./compiler/abstractinterpretation.jl:2418
_typeinf at ./compiler/typeinfer.jl:230
typeinf at ./compiler/typeinfer.jl:213
typeinf_edge at ./compiler/typeinfer.jl:895
abstract_call_method at ./compiler/abstractinterpretation.jl:601
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:161
abstract_call_known at ./compiler/abstractinterpretation.jl:1695
abstract_call at ./compiler/abstractinterpretation.jl:1767
abstract_call at ./compiler/abstractinterpretation.jl:1732
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1888
typeinf_local at ./compiler/abstractinterpretation.jl:2341
typeinf_nocycle at ./compiler/abstractinterpretation.jl:2418
_typeinf at ./compiler/typeinfer.jl:230
typeinf at ./compiler/typeinfer.jl:213
typeinf_edge at ./compiler/typeinfer.jl:895
abstract_call_method at ./compiler/abstractinterpretation.jl:601
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:161
abstract_call at ./compiler/abstractinterpretation.jl:1764
abstract_call at ./compiler/abstractinterpretation.jl:1732
abstract_eval_statement at ./compiler/abstractinterpretation.jl:1888
typeinf_local at ./compiler/abstractinterpretation.jl:2341
typeinf_nocycle at ./compiler/abstractinterpretation.jl:2418
_typeinf at ./compiler/typeinfer.jl:230
typeinf at ./compiler/typeinfer.jl:213
typeinf_ext at ./compiler/typeinfer.jl:986
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1019
typeinf_ext_toplevel at ./compiler/typeinfer.jl:1015
jfptr_typeinf_ext_toplevel_15682 at /home/chriselrod/Documents/languages/juliamaster/usr/lib/julia/sys.so (unknown line)
_jl_invoke at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2392 [inlined]
ijl_apply_generic at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2574 [inlined]
jl_apply at /home/chriselrod/Documents/languages/juliamaster/src/julia.h:1841 [inlined]
jl_type_infer at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:317
jl_generate_fptr_impl at /home/chriselrod/Documents/languages/juliamaster/src/jitlayers.cpp:344
jl_compile_method_internal at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2106
jl_compile_method_internal at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2050 [inlined]
_jl_invoke at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2384 [inlined]
ijl_apply_generic at /home/chriselrod/Documents/languages/juliamaster/src/gf.c:2574
jl_apply at /home/chriselrod/Documents/languages/juliamaster/src/julia.h:1841 [inlined]
do_call at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:126
eval_value at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:215
eval_stmt_value at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:166 [inlined]
eval_body at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:612
eval_body at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:522
eval_body at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:522
jl_interpret_toplevel_thunk at /home/chriselrod/Documents/languages/juliamaster/src/interpreter.c:750
macro expansion at /home/chriselrod/.julia/dev/LoopVectorization/test/miscellaneous.jl:1117
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:912
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:856
ijl_toplevel_eval at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:921 [inlined]
ijl_toplevel_eval_in at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
include_string at ./loading.jl:1349
_include at ./loading.jl:1409
include at ./client.jl:472
unknown function (ip: 0x7f07f01fe5e2)
macro expansion at ./timing.jl:267 [inlined]
macro expansion at /home/chriselrod/.julia/dev/LoopVectorization/test/grouptests.jl:73 [inlined]
macro expansion at ./timing.jl:267 [inlined]
macro expansion at /home/chriselrod/.julia/dev/LoopVectorization/test/grouptests.jl:70 [inlined]
macro expansion at /home/chriselrod/Documents/languages/juliamaster/usr/share/julia/stdlib/v1.9/Test/src/Test.jl:1385 [inlined]
macro expansion at /home/chriselrod/.julia/dev/LoopVectorization/test/grouptests.jl:9 [inlined]
top-level scope at ./timing.jl:267
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:903
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:856
ijl_toplevel_eval at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:921 [inlined]
ijl_toplevel_eval_in at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
include_string at ./loading.jl:1349
_include at ./loading.jl:1409
include at ./client.jl:472 [inlined]
top-level scope at /home/chriselrod/.julia/dev/LoopVectorization/test/runtests.jl:54
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:903
jl_toplevel_eval_flex at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:856
ijl_toplevel_eval at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:921 [inlined]
ijl_toplevel_eval_in at /home/chriselrod/Documents/languages/juliamaster/src/toplevel.c:971
eval at ./boot.jl:370 [inlined]
include_string at ./loading.jl:1349
_include at ./loading.jl:1409
include at ./Base.jl:428
jfptr_include_37280 at /home/chriselrod/Documents/languages/juliamaster/usr/lib/julia/sys.so (unknown line)
exec_options at ./client.jl:299
_start at ./client.jl:516
jfptr__start_32975 at /home/chriselrod/Documents/languages/juliamaster/usr/lib/julia/sys.so (unknown line)
jl_apply at /home/chriselrod/Documents/languages/juliamaster/src/julia.h:1841 [inlined]
true_main at /home/chriselrod/Documents/languages/juliamaster/src/jlapi.c:566
jl_repl_entrypoint at /home/chriselrod/Documents/languages/juliamaster/src/jlapi.c:710
main at /home/chriselrod/Documents/languages/juliamaster/cli/loader_exe.c:59
__libc_start_call_main at /lib64/libc.so.6 (unknown line)
__libc_start_main at /lib64/libc.so.6 (unknown line)
_start at /home/chriselrod/Documents/languages/juliamaster/usr/bin/julia (unknown line)
┌ Warning: #= /home/chriselrod/.julia/dev/LoopVectorization/test/copy.jl:182 =#:
│ `LoopVectorization.check_args` on your inputs failed; running fallback `@inbounds @fastmath` loop instead.
│ Use `warn_check_args=false`, e.g. `@turbo warn_check_args=false ...`, to disable this warning.
└ @ Main ~/.julia/dev/LoopVectorization/src/condense_loopset.jl:995
ERROR: LoadError: Some tests did not pass: 2489 passed, 0 failed, 1 errored, 0 broken.
in expression starting at /home/chriselrod/.julia/dev/LoopVectorization/test/grouptests.jl:7
in expression starting at /home/chriselrod/.julia/dev/LoopVectorization/test/runtests.jl:9
Julia Version 1.9.0-DEV.633
```
with the top of that trace caught in a loop/repeating endlessly.

**Comment by Tokazama** [comment]
Is that one an issue with the recent change to LayoutPointers?

**Comment by chriselrod** [comment]
> Is that one an issue with the recent change to LayoutPointers?

Yes. https://github.com/JuliaSIMD/LayoutPointers.jl/commit/327ebf0312463d38da256748c7af8664ca9413a7
I'm manually running more tests locally to make sure releasing all of these won't cause regressions.

**Comment by Tokazama** [comment]
> > Is that one an issue with the recent change to LayoutPointers?
> 
> Yes. [JuliaSIMD/LayoutPointers.jl@327ebf0](https://github.com/JuliaSIMD/LayoutPointers.jl/commit/327ebf0312463d38da256748c7af8664ca9413a7) I'm manually running more tests locally to make sure releasing all of these won't cause regressions.

Sorry. It looks like you've already taken care of it, but let me know if you need me to do anything.

**Review by chriselrod — APPROVED** [review]
There was a similar problem in CloseOpenIntervals, where it needed to support more integer types (in this case, fully `Integer` due to some LoopVectorization tests using `UpperBoundedInteger`s).
But LoopVectorization, StrideArrays, and Octavian tests passed locally.

**Comment by Tokazama** [comment]
I thought that `Int` was the fastest type for iteration because it matched the machine's bit size? Is there some other advantage to indexing with other bit types?

**Comment by chriselrod** [comment]
> I thought that `Int` was the fastest type for iteration because it matched the machine's bit size? Is there some other advantage to indexing with other bit types?

LoopVectorization assumes `UpperBoundedInteger`s are smaller than a statically known upper bound.
Octavian uses them: https://github.com/JuliaLinearAlgebra/Octavian.jl/blob/82a68e727f1ec2647664c801df661359c08c376a/src/macrokernels.jl#L41

**Comment by chriselrod** [comment]
But, yeah, generally `Int`/`UInt` should be "fastest".
Some operations on `Int32` may actually be faster on x86, because some `e` versions of instructions are 1 byte smaller than the `r` versions, reducing code size.

However, actually indexing requires 64 bit integers (except for some SIMD instructions like gather/scatter that can use 32 bit or 64 bit); cleaned a bit:
```asm
# @code_native debuginfo = :none syntax = :intel Base.unsafe_getindex(x, 3)
	mov	rax, qword ptr [rdi]
	vmovsd	xmm0, qword ptr [rax + 8*rsi - 8] # xmm0 = mem[0],zero
	ret
```
```asm
# @code_native debuginfo = :none syntax = :intel Base.unsafe_getindex(x, 3 % Int32)
	movsxd	rax, esi
	mov	rcx, qword ptr [rdi]
	vmovsd	xmm0, qword ptr [rcx + 8*rax - 8] # xmm0 = mem[0],zero
	ret
```
The 64 bit index version loads the pointer to the array `x`, and then indexes into it.
The 32 bit index version is the same, except it first extends the 32 bit index into a 64 bit index.

Obviously using 32 bits isn't going to decrease code size if it adds more code.

**Comment by chriselrod** [comment]
With respect to `FastRange`, it is to represent ranges of different types.

`FastRange` is used when you're doing something other than using integers to index, so it wants to represent the type of the original range. If you're SIMD-ing it, smaller types are of course faster.

**Comment by Tokazama** [comment]
Is there any benefit to using floats instead of integers for indexing?

**Comment by chriselrod** [comment]
> Is there any benefit to using floats instead of integers for indexing?

Ha ha, no, integers are definitely way better.
But JavaScript and R both default to floats.

**Comment by Tokazama** [comment]
Gotcha. Thank you for patiently answering my questions.

**Comment by chriselrod** [comment]
No problem.

Maybe "way" is an exaggeration, because converting floating point to integers isn't that expensive, but floating point cannot be used directly. Thus, it's nicer to avoid the need for conversion in the first place.

Also, consider that an Alder Lake performance core can do 5 integer adds/clock cycle, with a latency of a single clock cycle vs 2 floating point adds/clock cycle, with a latency of 3 cycles.

**Pull request merged**